File: ios/App/App/plugin/CapacitorWidgetPlugin.swift
import Foundation
import Capacitor
import WidgetKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorWidgetPlugin)
public class CapacitorWidgetPlugin: CAPPlugin {
    // private let implementation = CapacitorWidget()

    @objc func setItem(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key must be provided")
            return
        }
        
        guard let value = call.getString("value") else {
            call.reject("Value must be provided")
            return
        }
        
        guard let group = call.getString("group") else {
            call.reject("Group must be provided")
            return
        }
        
        if let userDefaults = UserDefaults.init(suiteName: group) {
            userDefaults.set(value, forKey: key)
            call.resolve(["results": true])
            return
        }
        
        call.reject("Could not set item")
    }
    
    @objc func getItem(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key must be provided")
            return
        }
        
        guard let group = call.getString("group") else {
            call.reject("Group must be provided")
            return
        }
        
        if let userDefaults = UserDefaults.init(suiteName: group) {
            let value = userDefaults.value(forKey: key)
            if value != nil {
                call.resolve(["results": value!])
                return
            }
        }
        
        call.reject("Could not get item")
    }
    
    @objc func removeItem(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key must be provided")
            return
        }
        
        guard let group = call.getString("group") else {
            call.reject("Group must be provided")
            return
        }
        
        if let userDefaults = UserDefaults.init(suiteName: group) {
            userDefaults.removeObject(forKey: key)
            call.resolve(["results": true])
            return
        }
        
        call.reject("Could not remove item")
    }
    
    @objc func reloadWidget(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            let widgetAppId = call.getString("widgetAppId")
            
            

            if (widgetAppId != nil) {
                WidgetCenter.shared.reloadTimelines(ofKind: widgetAppId!)
            } else {
                WidgetCenter.shared.reloadAllTimelines()
            }
            
            call.resolve(["results": true])
        } else {
            call.unavailable("Not available in iOS 13 or earlier.")
        }
    }
}

File: android/app/src/main/java/com/example/plugin/CapacitorWidgetPlugin.kt
package com.example.plugin

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "CapacitorWidget")
class CapacitorWidgetPlugin : Plugin() {

    private lateinit var implementation: CapacitorWidget

    override fun load() {
        implementation = CapacitorWidget(context)
    }

    @PluginMethod
    fun setItem(call: PluginCall) {
        val key = call.getString("key")
        if (key == null) {
            call.reject("Must provide key");
            return;
        }

        val value = call.getString("value")

        if (value == null) {
            call.reject("Must provide value");
            return;
        }

        val group = call.getString("group")

        if (group == null) {
            call.reject("Must provide group");
            return;
        }

        val ret = JSObject()
        ret.put("results", implementation.setItem(key, value, group))
        call.resolve(ret)
    }

    @PluginMethod
    fun getItem(call: PluginCall) {
        val key = call.getString("key")
        if (key == null) {
            call.reject("Must provide key");
            return;
        }

        val group = call.getString("group")

        if (group == null) {
            call.reject("Must provide group");
            return;
        }

        val ret = JSObject()
        ret.put("results", implementation.getItem(key, group))
        call.resolve(ret)
    }

    @PluginMethod
    fun removeItem(call: PluginCall) {
        val key = call.getString("key")
        if (key == null) {
            call.reject("Must provide key");
            return;
        }

        val group = call.getString("group")

        if (group == null) {
            call.reject("Must provide group");
            return;
        }

        val ret = JSObject()
        ret.put("results", implementation.removeItem(key, group))
        call.resolve(ret)
    }

    @PluginMethod
    fun reloadWidget(call: PluginCall) {
        val widgetAppId = call.getString("widgetAppId")

        val ret = JSObject()
        ret.put("results", implementation.reloadWidget())
        call.resolve()
    }

}